File: src/skills/getRecentSearches.js
import { query } from '../db/pool.js';

const MAX_HISTORY = 5;

/**
 * Retrieve the most recent flight searches for a given route.
 * Used to provide historical price context to the sales opportunity analysis.
 *
 * @param {object} params
 * @param {string} params.origin       IATA origin code
 * @param {string} params.destination  IATA destination code
 * @param {number} [params.limit]      Max records to return (default 5)
 * @returns {Promise<Array<{price: number|null, departureDate: string, searchedAt: string}>>}
 */
export default async function getRecentSearches({
  origin,
  destination,
  limit = MAX_HISTORY,
}) {
  if (!origin || !destination) return [];

  try {
    const result = await query(
      `SELECT
         cheapest_flight->>'price' AS price,
         departure_date            AS "departureDate",
         created_at                AS "searchedAt"
       FROM flight_searches
       WHERE origin = $1 AND destination = $2
       ORDER BY created_at DESC
       LIMIT $3`,
      [origin.toUpperCase(), destination.toUpperCase(), limit],
    );

    return result.rows.map((row) => ({
      price: row.price != null ? Number(row.price) : null,
      departureDate: row.departureDate,
      searchedAt: row.searchedAt,
    }));
  } catch (err) {
    console.error('[getRecentSearches] Failed:', err.message);
    return [];
  }
}

File: app/skills/get_recent_searches.py
"""
skills/get_recent_searches.py — Historical price data for trend analysis.
"""

from app.db.pool import query

MAX_HISTORY = 5


def get_recent_searches(origin: str, destination: str, limit: int = MAX_HISTORY) -> list[dict]:
    """Return last `limit` flight search prices for the given route."""
    if not origin or not destination:
        return []

    try:
        rows = query(
            """
            SELECT
                cheapest_flight->>'price' AS price,
                departure_date            AS departure_date,
                created_at                AS searched_at
            FROM flight_searches
            WHERE origin = %s AND destination = %s
            ORDER BY created_at DESC
            LIMIT %s
            """,
            (origin.upper(), destination.upper(), limit),
        )
        return [
            {
                "price": float(r["price"]) if r.get("price") is not None else None,
                "departure_date": r.get("departure_date"),
                "searched_at": r.get("searched_at"),
            }
            for r in rows
        ]
    except Exception as err:
        print(f"[get_recent_searches] Failed: {err}")
        return []
